README.md
[![NPM Version](https://badge.fury.io/js/driver-license-validator.svg)](https://badge.fury.io/js/driver-license-validator)
[![CI](https://github.com/justinlettau/driver-license-validator/workflows/CI/badge.svg)](https://github.com/justinlettau/driver-license-validator/actions)
[![codecov](https://codecov.io/gh/justinlettau/driver-license-validator/branch/master/graph/badge.svg)](https://codecov.io/gh/justinlettau/driver-license-validator)

# Driver License Validator

Validate driver license numbers.
Formats based on [https://ntsi.com/drivers-license-format/](https://ntsi.com/drivers-license-format/).

# Table of Contents

- [Installation](#installation)
- [Usage](#usage)
- [Development](#development)

# Installation

```bash
npm install driver-license-validator --save
```

# Usage

```js
import { getMatches, isValid } from 'driver-license-validator';

isValid('A1234567');
// => true

isValid('invalid_dl');
// => false

getMatches('A1234567');
// => [
//   { description: '1 letter followed by 1-8 numbers', state: 'AZ' },
//   { description: '1 letter followed by 7 numbers', state: 'CA' },
//   { description: '1 letter followed by 5-9 numbers', state: 'MO' },
//   { description: '1 letter followed by 6-8 numbers', state: 'NE' },
//   { description: '1 letter followed by 7 numbers', state: 'NY' },
//   { description: '1 letter followed by 4-8 numbers', state: 'OH' }
// ]

getMatches('A1234567', { states: 'NY' });
// => [
//   { description: '1 letter followed by 7 numbers', state: 'NY' }
// ]

getMatches('A1234567', { states: ['NY', 'OH'] });
// => [
//   { description: '1 letter followed by 7 numbers', state: 'NY' },
//   { description: '1 letter followed by 4-8 numbers', state: 'OH' }
// ]

getMatches('A123456789', { country: 'CA' });
// => [
//   { description: '1 letter followed by 9 numbers', state: 'NL' }
// ]

getMatches('invalid_dl');
// => null
```

Options:

| Option       | Type                   | Description                    | Default |
| ------------ | ---------------------- | ------------------------------ | ------- |
| `country`    | `string`               | Country to use.                | `US`    |
| `states`     | `string` or `string[]` | State(s) to filter results to. | n/a     |
| `ignoreCase` | `boolean`              | True to ignore case.           | `false` |

# Development

```
npm install
npm run build
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [3.1.1](https://github.com/justinlettau/driver-license-validator/compare/v3.1.0...v3.1.1) (2022-01-08)


### Bug Fixes

* esm import extensions ([d8e10b0](https://github.com/justinlettau/driver-license-validator/commit/d8e10b07980582b28e0e3f3695184e3402110704))

## [3.1.0](https://github.com/justinlettau/driver-license-validator/compare/v3.0.0...v3.1.0) (2021-12-13)


### Features

* add MO format ([#45](https://github.com/justinlettau/driver-license-validator/issues/45)) ([0d240c9](https://github.com/justinlettau/driver-license-validator/commit/0d240c90cadbad4b7ce82ca44e6c0fd401b86f9d)), closes [#44](https://github.com/justinlettau/driver-license-validator/issues/44)

## [3.0.0](https://github.com/justinlettau/driver-license-validator/compare/v2.0.2...v3.0.0) (2021-10-28)


### ⚠ BREAKING CHANGES

* This package is now pure ESM

### Features

* esm package ([6a0916e](https://github.com/justinlettau/driver-license-validator/commit/6a0916ed00cc6a14543385fccff0e5de138242c8))

### [2.0.2](https://github.com/justinlettau/driver-license-validator/compare/v2.0.1...v2.0.2) (2021-05-19)


### Bug Fixes

* update us regex ([93101b7](https://github.com/justinlettau/driver-license-validator/commit/93101b70281e13e2d2df4473a0cf55f4b5609e83)), closes [#25](https://github.com/justinlettau/driver-license-validator/issues/25)

### [2.0.1](https://github.com/justinlettau/driver-license-validator/compare/v2.0.0...v2.0.1) (2021-03-23)

## [2.0.0](https://github.com/justinlettau/driver-license-validator/compare/v1.0.0...v2.0.0) (2020-11-23)


### ⚠ BREAKING CHANGES

* rename `validate` method to `getMatches`
* `validate` signature changed

### Features

* add CA country support ([83d7268](https://github.com/justinlettau/driver-license-validator/commit/83d72685c54df38ec8967f2e80b75eaa430fd86b))
* add ignore case option ([67ef69f](https://github.com/justinlettau/driver-license-validator/commit/67ef69f5c4895bb36ce143aba0800c09b406e3b0))
* add isValid method ([905038d](https://github.com/justinlettau/driver-license-validator/commit/905038d8190eab81a237d55b0bb0faf2d1bdc050))
* rename validate method ([ba81269](https://github.com/justinlettau/driver-license-validator/commit/ba812699fe4f47965b90c4f1b8a65ecfbd6d24ff))


### Bug Fixes

* AZ and VA regex ([fac2226](https://github.com/justinlettau/driver-license-validator/commit/fac222650655c046bd484c0814f6d153dbaa1aef))

# 1.0.0 (2019-04-16)


src/index.spec.ts
import { isValid, getMatches } from './index';

const state_fixture: StateTestFormat = {
  AL: [
    {
      description: '7-8 numbers',
      sample: ['1234567', '12345678'],
    },
  ],
  AK: [
    {
      description: '7 numbers',
      sample: ['1234567'],
    },
  ],
  AZ: [
    {
      description: '1 letter followed by 8-9 numbers',
      sample: ['A12345678', 'A123456789'],
    },
    {
      description: '2 letters followed by 2-5 numbers',
      sample: ['AB12', 'AB123', 'AB1234', 'AB12345'],
    },
    {
      description: '9 numbers',
      sample: ['123456789'],
    },
  ],
  AR: [
    {
      description: '9 numbers',
      sample: ['123456789'],
    },
  ],
  CA: [
    {
      description: '1 letter followed by 7 numbers',
      sample: ['A1234567'],
    },
  ],
  CO: [
    {
      description: '9 numbers',
      sample: ['123456789'],
    },
    {
      description: '1 letter followed by 3-6 numbers',
      sample: ['A123', 'A1234', 'A12345', 'A123456'],
    },
    {
      description: '2 letters followed by 2-5 numbers',
      sample: ['AB12', 'AB123', 'AB1234', 'AB12345'],
    },
  ],
  CT: [
    {
      description: '9 numbers',
      sample: ['123456789'],
    },
  ],
  DE: [
    {
      description: '7 numbers',
      sample: ['1234567'],
    },
  ],
  DC: [
    {
      description: '7 numbers',
      sample: ['1234567'],
    },
    {
      description: '9 numbers',
      sample: ['123456789'],
    },
  ],
  FL: [
    {
      description: '1 letter followed by 12 numbers',
      sample: ['A123456789012'],
    },
  ],
  GA: [
    {
      description: '7-9 numbers',
      sample: ['1234567', '12345678', '123456789'],
    },
  ],
  HI: [
    {
      description: '1 letter followed by 8 numbers',
      sample: ['A12345678'],
    },
    {
      description: '9 numbers',
      sample: ['123456789'],
    },
  ],
  ID: [
    {
      description: '2 letters followed by 6 numbers followed by 1 letter',
      sample: ['AA123456Z'],
    },
    {
      description: '9 Numbers',
      sample: ['123456789'],
    },
  ],
  IL: [
    {
      description: '1 letter followed by 11-12 numbers',
      sample: ['A123456789012', 'A12345678901'],
    },
  ],
  IN: [
    {
      description: '1 letter followed by 9 numbers',
      sample: ['A123456789'],
    },
    {
      description: '9-10 numbers',
      sample: ['123456789', '1234567890'],
    },
  ],
  IA: [
    {
      description: '9 numbers',
      sample: ['123456789'],
    },
    {
      description: '3 numbers followed by 2 letters followed by 4 numbers',
      sample: ['123AA1234'],
    },
  ],
  KS: [
    {
      description: '1 letter then 1 number then 1 letter then 1 number then 1 letter',
      sample: ['A1B2C'],
    },
    {
      description: '1 letter followed by 8 numbers',
      sample: ['A12345678'],
    },
    {
      description: '9 numbers',
      sample: ['123456789'],
    },
  ],
  KY: [
    {
      description: '1 letter followed by 8-9 numbers',
      sample: ['A12345678', 'A123456789'],
    },
    {
      description: '9 numbers',
      sample: ['123456789'],
    },
  ],
  LA: [
    {
      description: '9 numbers',
      sample: ['123456789'],
    },
  ],
  ME: [
    {
      description: '7 numbers',
      sample: ['1234567'],
    },
    {
      description: '7 numbers followed by 1 letter',
      sample: ['1234567A'],
    },
    {
      description: '8 numbers',
      sample: ['12345678'],
    },
  ],
  MD: [
    {
      description: '1 letter followed by 12 numbers',
      sample: ['A123456789012'],
    }
  ],
  MA: [
    {
      description: '1 letter followed by 8 numbers',
      sample: ['A12345678'],
    },
    {
      description: '9 numbers',
      sample: ['123456789'],
    },
  ],
  MI: [
    {
      description: '1 letter followed by 10 numbers',
      sample: ['A1234567890'],
    },
    {
      description: '1 letter followed by 12 numbers',
      sample: ['A123456789012'],
    },
  ],
  MN: [
    {
      description: '1 letter followed by 12 numbers',
      sample: ['A123456789012'],
    },
  ],
  MS: [
    {
      description: '9 numbers',
      sample: ['123456789'],
    },
  ],
  MO: [
    {
      description: '1 letter followed by 5-9 numbers',
      sample: ['A12345', 'A123456', 'A1234567', 'A12345678', 'A123456789'],
    },
    {
      description: '1 letter followed by 6 numbers followed by "R"',
      sample: ['A123456R'],
    },
    {
      description: '8 numbers followed by 2 letters',
      sample: ['12345678AA'],
    },
    {
      description: '9 numbers followed by 1 letter',
      sample: ['123456789A'],
    },
    {
      description: '9 numbers',
      sample: ['123456789'],
    },
    {
      description: '3 numbers followed by 1 letter followed by 6 numbers',
      sample: ['123A456789']
    },
  ],
  MT: [
    {
      description: '1 letter followed by 8 numbers',
      sample: ['A12345678'],
    },
    {
      description: '13 numbers',
      sample: ['1234567890123'],
    },
    {
      description: '9 numbers',
      sample: ['123456789'],
    },
    {
      description: '14 numbers',
      sample: ['12345678901234'],
    },
  ],
  NE: [
    {
      description: '1 letter followed by 6-8 numbers',
      sample: ['A123456', 'A1234567', 'A12345678'],
    },
  ],
  NV: [
    {
      description: '9-10 numbers',
      sample: ['123456789', '1234567890'],
    },
    {
      description: '12 numbers',
      sample: ['123456789012'],
    },
    {
      description: '"X" followed by 8 numbers',
      sample: ['X12345678'],
    },
  ],
  NH: [
    {
      description: '2 numbers followed by 3 letters followed by 5 numbers',
      sample: ['12AAA12345'],
    },
  ],
  NJ: [
    {
      description: '1 letter followed by 14 numbers',
      sample: ['C63812250002932'],
    },
  ],
  NM: [
    {
      description: '8-9 numbers',
      sample: ['12345678', '123456789'],
    },
  ],
  NY: [
    {
      description: '1 letter followed by 7 numbers',
      sample: ['A1234567'],
    },
    {
      description: '1 letter followed by 18 numbers',
      sample: ['A123456789012345678'],
    },
    {
      description: '8-9 numbers',
      sample: ['12345678', '123456789'],
    },
    {
      description: '16 numbers',
      sample: ['1234567890123456'],
    },
    {
      description: '8 letters',
      sample: ['ABCDEFGH'],
    },
  ],
  NC: [
    {
      description: '1-12 numbers',
      sample: [
        '1',
        '12',
        '123',
        '1234',
        '12345',
        '123456',
        '1234567',
        '12345678',
        '123456789',
        '1234567890',
        '12345678901',
        '123456789012',
      ],
    },
  ],
  ND: [
    {
      description: '3 letters followed by 6 numbers',
      sample: ['SHA942996'],
    },
    {
      description: '9 numbers',
      sample: ['123456789'],
    },
  ],
  OH: [
    {
      description: '1 letter followed by 4-8 numbers',
      sample: ['A1234', 'A12345', 'A123456', 'A1234567', 'A12345678'],
    },
    {
      description: '2 letters followed by 3-7 numbers',
      sample: ['AA123', 'AA1234', 'AA12345', 'AA123456', 'AA1234567'],
    },
    {
      description: '8 numbers',
      sample: ['12345678'],
    },
  ],
  OK: [
    {
      description: '1 letter followed by 9 numbers',
      sample: ['A123456789'],
    },
    {
      description: '9 numbers',
      sample: ['123456789'],
    },
  ],
  OR: [
    {
      description: '1-9 numbers',
      sample: ['1', '12', '123', '1234', '12345', '123456', '1234567', '12345678', '123456789'],
    },
  ],
  PA: [
    {
      description: '8 numbers',
      sample: ['12345678'],
    },
  ],
  RI: [
    {
      description: '7 numbers',
      sample: ['1234567'],
    },
    {
      description: '1 letter followed by 6 numbers',
      sample: ['A123456'],
    },
  ],
  SC: [
    {
      description: '5-11 numbers',
      sample: ['12345', '123456', '1234567', '12345678', '123456789', '1234567890', '12345678901'],
    },
  ],
  SD: [
    {
      description: '6-10 numbers',
      sample: ['123456', '1234567', '12345678', '123456789', '1234567890'],
    },
    {
      description: '12 numbers',
      sample: ['123456789012'],
    },
  ],
  TN: [
    {
      description: '7-9 numbers',
      sample: ['1234567', '12345678', '123456789'],
    },
  ],
  TX: [
    {
      description: '7-8 numbers',
      sample: ['1234567', '44505904'],
    },
  ],
  UT: [
    {
      description: '4-10 numbers',
      sample: ['1234', '12345', '123456', '1234567', '12345678', '123456789', '1234567890'],
    },
  ],
  VT: [
    {
      description: '8 numbers',
      sample: ['12345678'],
    },
    {
      description: '7 numbers followed by "A"',
      sample: ['1234567A'],
    },
  ],
  VA: [
    {
      description: '1 letter followed by 8-11 numbers',
      sample: ['A12345678', 'A123456789', 'A1234567890', 'A12345678901'],
    },
    {
      description: '9 numbers',
      sample: ['123456789'],
    },
  ],
  WA: [
    {
      description: '1-7 letters followed by any combination of letters, numbers, or "*" for a total of 12 characters',
      sample: [
        'A00000000000',
        'AB0000000000',
        'ABC000000000',
        'ABCD00000000',
        'ABCDE0000000',
        'ABCDEF000000',
        'ABCDEFG00000',
        'ABCDEFG*0000',
        'ABCDEFG**000',
        'ABCDEFG***00',
        'ABCDEFG****0',
        'ABCDEFG*****',
        'ABCDEFGA****',
        'ABCDEFGAB***',
        'ABCDEFGABC**',
        'ABCDEFGABCD*',
        'ABCDEFGABCDE',
      ],
    },
  ],
  WV: [
    {
      description: '7 numbers',
      sample: ['1234567'],
    },
    {
      description: '1-2 letters followed by 5-6 numbers',
      sample: ['A12345', 'A123456', 'AA12345', 'AA123456'],
    },
  ],
  WI: [
    {
      description: '1 letter followed by 13 numbers',
      sample: ['A1234567890123'],
    },
  ],
  WY: [
    {
      description: '9-10 numbers',
      sample: ['123456789', '1234567890'],
    },
  ],
};

describe('isValid method', () => {
  it('should return false when no matches found', () => {
    const result = isValid('invalid_dl');
    expect(result).toBe(false);
  });

  it('should return true when matches found', () => {
    const result = isValid('A1234567');
    expect(result).toBe(true);
  });
});

describe('getMatches method', () => {
  it('should return null when no match', () => {
    const result = getMatches('invalid_dl');
    expect(result).toBe(null);
  });

  it('should return all US matches', () => {
    const result = getMatches('A1234567');
    expect(result.length).toBe(5);
  });

  for (const state of Object.keys(state_fixture)) {
    describe(`Testing ${state} patterns`, function () {
      for (const pattern of state_fixture[state]) {
        describe(`${pattern.description}`, () => {
          for (const n of pattern.sample) {
            it(`${n} should match`, () => {
              const result = getMatches(n, { states: state });
              expect(result.length).toBe(1);
            });
          }
        });
      }
    });
  }

  it('invalid state should throw', () => {
    expect( () => getMatches( 'a1234567', { states: 'ZA' } )).toThrow();
  });

  it('states array should throw', () => {
    expect( () => getMatches('a1234567', { states: [ 'TX', 'AL' ] } )).toThrow();
  });

  it('should return all CA matches', () => {
    const result = getMatches('A123456789', { country: 'CA' });
    expect(result.length).toBe(1);
  });

  it('should return state US matches case insensitive', () => {
    const result = getMatches('ab1234', { ignoreCase: true });
    expect(result.length).toBe(3);
  });
});

export interface StateTestFormat {
  [index: string]: DriverLicenseTestFormat[];
}

export interface DriverLicenseTestFormat {
  description: string;
  sample: string | string[];
}


src/index.ts
import { CountryFormats, ValidateOptions, ValidationMatch } from './interfaces.js';
import { CA_DL } from './regex/ca-dl.js';
import { US_DL } from './regex/us-dl.js';

/**
 * Check if a driver license number matches any format.
 *
 * @param dl Driver license number.
 * @param options Optional configuration options.
 */
export function isValid(dl: string, options: ValidateOptions = {}) {
  const result = getMatches(dl, options);

  return result?.length > 0;
}

/**
 * Get all matching formats for a driver license number.
 *
 * @param dl Driver license number.
 * @param options Optional configuration options.
 */
export function getMatches(dl: string, options: ValidateOptions = {}) {
  const results: ValidationMatch[] = [];
  let formats: CountryFormats;
  let states: string[];

  switch (options.country) {
    case 'CA':
      formats = CA_DL;
      break;
    case 'US':
    default:
      formats = US_DL;
      break;
  }

  if (!options.states) {
    states = Object.keys(formats);
  } else if (!Array.isArray(options.states)) {
    states = [options.states];
  }

  states.forEach((state) => {
    const info = formats[state];

    if (!info) {
      throw new Error(`Could not find state "${state}"!`);
    }

    info.forEach((item) => {
      const { regex, description } = item;
      const pattern = options.ignoreCase ? new RegExp(regex, 'i') : regex;

      if (pattern.test(dl)) {
        results.push({
          description,
          state,
        });
      }
    });
  });

  return results.length ? results : null;
}


src/regex/ca-dl.ts
import { CountryFormats } from '../interfaces.js';

/**
 * Canada driver license formats.
 */
export const CA_DL: CountryFormats = {
  AB: [
    {
      regex: /^[0-9]{9}$/,
      description: '9 numbers',
    },
  ],
  BC: [
    {
      regex: /^[0-9]{7}$/,
      description: '7 numbers',
    },
  ],
  MB: [
    {
      regex: /^[A-Z0-9]{12}$/,
      description: 'any combination of letters or numbers for a total of 12 characters',
    },
  ],
  NB: [
    {
      regex: /^[0-9]{5,7}$/,
      description: '5-7 numbers',
    },
  ],
  NL: [
    {
      regex: /^[A-Z]{1}[0-9]{9}$/,
      description: '1 letter followed by 9 numbers',
    },
  ],
  NT: [
    {
      regex: /^[0-9]{6}$/,
      description: '6 numbers',
    },
  ],
  NS: [
    {
      regex: /^[A-Z]{5}[0-9]{9}$/,
      description: '5 letters followed by 9 numbers',
    },
  ],
  NU: [
    {
      regex: /^[0-9]{9}$/,
      description: '9 numbers',
    },
  ],
  ON: [
    {
      regex: /^[A-Z]{1}[0-9]{14}$/,
      description: '1 letter followed by 14 numbers',
    },
  ],
  PE: [
    {
      regex: /^[0-9]{5,6}$/,
      description: '5-6 numbers',
    },
  ],
  QC: [
    {
      regex: /^[A-Z]{1}[0-9]{12}$/,
      description: '1 letter followed by 12 numbers',
    },
  ],
  SK: [
    {
      regex: /^[0-9]{8}$/,
      description: '8 numbers',
    },
  ],
  YT: [
    {
      regex: /^[0-9]{10}$/,
      description: '10 numbers',
    },
  ],
};


src/regex/us-dl.ts
import { CountryFormats } from '../interfaces.js';

/**
 * USA driver license formats.
 */
export const US_DL: CountryFormats = {
  AL: [
    {
      regex: /^[0-9]{1,8}$/,
      description: '1-8 numbers',
    },
  ],
  AK: [
    {
      regex: /^[0-9]{1,7}$/,
      description: '1-7 numbers',
    },
  ],
  AZ: [
    {
      regex: /^[A-Z]{1}[0-9]{8,9}$/,
      description: '1 letter followed by 8-9 numbers',
    },
    {
      regex: /^[A-Z]{2}[0-9]{2,5}$/,
      description: '2 letters followed by 2-5 numbers',
    },
    {
      regex: /^[0-9]{9}$/,
      description: '9 numbers',
    },
  ],
  AR: [
    {
      regex: /^[0-9]{4,9}$/,
      description: '4-9 numbers',
    },
  ],
  CA: [
    {
      regex: /^[A-Z]{1}[0-9]{7}$/,
      description: '1 letter followed by 7 numbers',
    },
  ],
  CO: [
    {
      regex: /^[0-9]{9}$/,
      description: '9 numbers',
    },
    {
      regex: /^[A-Z]{1}[0-9]{3,6}$/,
      description: '1 letter followed by 3-6 numbers',
    },
    {
      regex: /^[A-Z]{2}[0-9]{2,5}$/,
      description: '2 letters followed by 2-5 numbers',
    },
  ],
  CT: [
    {
      regex: /^[0-9]{9}$/,
      description: '9 numbers',
    },
  ],
  DE: [
    {
      regex: /^[0-9]{1,7}$/,
      description: '1-7 numbers',
    },
  ],
  DC: [
    {
      regex: /^[0-9]{7}$/,
      description: '7 numbers',
    },
    {
      regex: /^[0-9]{9}$/,
      description: '9 numbers',
    },
  ],
  FL: [
    {
      regex: /^[A-Z]{1}[0-9]{12}$/,
      description: '1 letter followed by 12 numbers',
    },
  ],
  GA: [
    {
      regex: /^[0-9]{7,9}$/,
      description: '7-9 numbers',
    },
  ],
  HI: [
    {
      regex: /^[A-Z]{1}[0-9]{8}$/,
      description: '1 letter followed by 8 numbers',
    },
    {
      regex: /^[0-9]{9}$/,
      description: '9 numbers',
    },
  ],
  ID: [
    {
      regex: /^[A-Z]{2}[0-9]{6}[A-Z]{1}$/,
      description: '2 letters followed by 6 numbers followed by 1 letter',
    },
    {
      regex: /^[0-9]{9}$/,
      description: '9 numbers',
    },
  ],
  IL: [
    {
      regex: /^[A-Z]{1}[0-9]{11,12}$/,
      description: '1 letter followed by 11-12 numbers',
    },
  ],
  IN: [
    {
      regex: /^[A-Z]{1}[0-9]{9}$/,
      description: '1 letter followed by 9 numbers',
    },
    {
      regex: /^[0-9]{9,10}$/,
      description: '9-10 numbers',
    },
  ],
  IA: [
    {
      regex: /^[0-9]{9}$/,
      description: '9 numbers',
    },
    {
      regex: /^[0-9]{3}[A-Z]{2}[0-9]{4}$/,
      description: '3 numbers followed by 2 letters followed by 4 numbers',
    },
  ],
  KS: [
    {
      regex: /^([A-Z]{1}[0-9]{1}){2}[A-Z]{1}$/,
      description: '1 letter then 1 number then 1 letter then 1 number then 1 letter',
    },
    {
      regex: /^[A-Z]{1}[0-9]{8}$/,
      description: '1 letter followed by 8 numbers',
    },
    {
      regex: /^[0-9]{9}$/,
      description: '9 numbers',
    },
  ],
  KY: [
    {
      regex: /^[A-Z]{1}[0-9]{8,9}$/,
      description: '1 letter followed by 8-9 numbers',
    },
    {
      regex: /^[0-9]{9}$/,
      description: '9 numbers',
    },
  ],
  LA: [
    {
      regex: /^[0-9]{1,9}$/,
      description: '1-9 numbers',
    },
  ],
  ME: [
    {
      regex: /^[0-9]{7}$/,
      description: '7 numbers',
    },
    {
      regex: /^[0-9]{7}[A-Z]{1}$/,
      description: '7 numbers followed by 1 letter',
    },
    {
      regex: /^[0-9]{8}$/,
      description: '8 numbers',
    },
  ],
  MD: [
    {
      regex: /^[A-Z]{1}[0-9]{12}$/,
      description: '1 letter followed by 12 numbers',
    },
  ],
  MA: [
    {
      regex: /^[A-Z]{1}[0-9]{8}$/,
      description: '1 letter followed by 8 numbers',
    },
    {
      regex: /^[0-9]{9}$/,
      description: '9 numbers',
    },
  ],
  MI: [
    {
      regex: /^[A-Z]{1}[0-9]{10}$/,
      description: '1 letter followed by 10 numbers',
    },
    {
      regex: /^[A-Z]{1}[0-9]{12}$/,
      description: '1 letter followed by 12 numbers',
    },
  ],
  MN: [
    {
      regex: /^[A-Z]{1}[0-9]{12}$/,
      description: '1 letter followed by 12 numbers',
    },
  ],
  MS: [
    {
      regex: /^[0-9]{9}$/,
      description: '9 numbers',
    },
  ],
  MO: [
    {
      regex: /^[A-Z]{1}[0-9]{5,9}$/,
      description: '1 letter followed by 5-9 numbers',
    },
    {
      regex: /^[A-Z]{1}[0-9]{6}[R]{1}$/,
      description: '1 letter followed by 6 numbers followed by "R"',
    },
    {
      regex: /^[0-9]{8}[A-Z]{2}$/,
      description: '8 numbers followed by 2 letters',
    },
    {
      regex: /^[0-9]{9}[A-Z]{1}$/,
      description: '9 numbers followed by 1 letter',
    },
    {
      regex: /^[0-9]{9}$/,
      description: '9 numbers',
    },
    {
      regex: /^[0-9]{3}[A-Z]{1}[0-9]{6}$/,
      description: '3 numbers followed by 1 letter followed by 6 numbers',
    },
  ],
  MT: [
    {
      regex: /^[A-Z]{1}[0-9]{8}$/,
      description: '1 letter followed by 8 numbers',
    },
    {
      regex: /^[0-9]{13}$/,
      description: '13 numbers',
    },
    {
      regex: /^[0-9]{9}$/,
      description: '9 numbers',
    },
    {
      regex: /^[0-9]{14}$/,
      description: '14 numbers',
    },
  ],
  NE: [
    {
      regex: /^[A-Z]{1}[0-9]{6,8}$/,
      description: '1 letter followed by 6-8 numbers',
    },
  ],
  NV: [
    {
      regex: /^[0-9]{9,10}$/,
      description: '9-10 numbers',
    },
    {
      regex: /^[0-9]{12}$/,
      description: '12 numbers',
    },
    {
      regex: /^[X]{1}[0-9]{8}$/,
      description: '"X" followed by 8 numbers',
    },
  ],
  NH: [
    {
      regex: /^[0-9]{2}[A-Z]{3}[0-9]{5}$/,
      description: '2 numbers followed by 3 letters followed by 5 numbers',
    },
  ],
  NJ: [
    {
      regex: /^[A-Z]{1}[0-9]{14}$/,
      description: '1 letter followed by 14 numbers',
    },
  ],
  NM: [
    {
      regex: /^[0-9]{8,9}$/,
      description: '8-9 numbers',
    },
  ],
  NY: [
    {
      regex: /^[A-Z]{1}[0-9]{7}$/,
      description: '1 letter followed by 7 numbers',
    },
    {
      regex: /^[A-Z]{1}[0-9]{18}$/,
      description: '1 letter followed by 18 numbers',
    },
    {
      regex: /^[0-9]{8,9}$/,
      description: '8-9 numbers',
    },
    {
      regex: /^[0-9]{16}$/,
      description: '16 numbers',
    },
    {
      regex: /^[A-Z]{8}$/,
      description: '8 letters',
    },
  ],
  NC: [
    {
      regex: /^[0-9]{1,12}$/,
      description: '1-12 numbers',
    },
  ],
  ND: [
    {
      regex: /^[A-Z]{3}[0-9]{6}$/,
      description: '3 letters followed by 6 numbers',
    },
    {
      regex: /^[0-9]{9}$/,
      description: '9 numbers',
    },
  ],
  OH: [
    {
      regex: /^[A-Z]{1}[0-9]{4,8}$/,
      description: '1 letter followed by 4-8 numbers',
    },
    {
      regex: /^[A-Z]{2}[0-9]{3,7}$/,
      description: '2 letters followed by 3-7 numbers',
    },
    {
      regex: /^[0-9]{8}$/,
      description: '8 numbers',
    },
  ],
  OK: [
    {
      regex: /^[A-Z]{1}[0-9]{9}$/,
      description: '1 letter followed by 9 numbers',
    },
    {
      regex: /^[0-9]{9}$/,
      description: '9 numbers',
    },
  ],
  OR: [
    {
      regex: /^[0-9]{1,9}$/,
      description: '1-9 numbers',
    },
  ],
  PA: [
    {
      regex: /^[0-9]{8}$/,
      description: '8 numbers',
    },
  ],
  RI: [
    {
      regex: /^[0-9]{7}$/,
      description: '7 numbers',
    },
    {
      regex: /^[A-Z]{1}[0-9]{6}$/,
      description: '1 letter followed by 6 numbers',
    },
  ],
  SC: [
    {
      regex: /^[0-9]{5,11}$/,
      description: '5-11 numbers',
    },
  ],
  SD: [
    {
      regex: /^[0-9]{6,10}$/,
      description: '6-10 numbers',
    },
    {
      regex: /^[0-9]{12}$/,
      description: '12 numbers',
    },
  ],
  TN: [
    {
      regex: /^[0-9]{7,9}$/,
      description: '7-9 numbers',
    },
  ],
  TX: [
    {
      regex: /^[0-9]{7,8}$/,
      description: '7-8 numbers',
    },
  ],
  UT: [
    {
      regex: /^[0-9]{4,10}$/,
      description: '4-10 numbers',
    },
  ],
  VT: [
    {
      regex: /^[0-9]{8}$/,
      description: '8 numbers',
    },
    {
      regex: /^[0-9]{7}[A]$/,
      description: '7 numbers followed by "A"',
    },
  ],
  VA: [
    {
      regex: /^[A-Z]{1}[0-9]{8,11}$/,
      description: '1 letter followed by 8-11 numbers',
    },
    {
      regex: /^[0-9]{9}$/,
      description: '9 numbers',
    },
  ],
  WA: [
    {
      regex: /^(?=.{12}$)[A-Z]{1,7}[A-Z0-9\\*]{4,11}$/,
      description: '1-7 letters followed by any combination of letters, numbers, or "*" for a total of 12 characters',
    },
  ],
  WV: [
    {
      regex: /^[0-9]{7}$/,
      description: '7 numbers',
    },
    {
      regex: /^[A-Z]{1,2}[0-9]{5,6}$/,
      description: '1-2 letters followed by 5-6 numbers',
    },
  ],
  WI: [
    {
      regex: /^[A-Z]{1}[0-9]{13}$/,
      description: '1 letter followed by 13 numbers',
    },
  ],
  WY: [
    {
      regex: /^[0-9]{9,10}$/,
      description: '9-10 numbers',
    },
  ],
};
